examples/connectivity/plot_ica_neurovault.py

```diff
@@ -0,0 +1,244 @@
+"""
+NeuroVault cross-study ICA maps.
+================================
+
+This example shows how to download statistical maps from
+NeuroVault, label them with NeuroSynth terms,
+and compute ICA components across all the maps.
+
+See :func:`nilearn.datasets.fetch_neurovault` documentation for more details.
+"""
+# Author: Ben Cipollini
+# License: BSD
+# Ported from code authored by Chris Filo Gorgolewski, Gael Varoquaux
+# https://github.com/NeuroVault/neurovault_analysis
+
+import json
+import io
+import os
+import warnings
+warnings.simplefilter('error', RuntimeWarning)
+
+import numpy as np
+from joblib import Memory
+from matplotlib import pyplot as plt
+from scipy import stats
+from sklearn.decomposition import FastICA
+from sklearn.feature_extraction import DictVectorizer
+
+import nibabel as nib
+from nilearn import datasets
+from nilearn.datasets.utils import _fetch_files, _get_dataset_dir
+from nilearn.image import resample_img
+from nilearn.input_data import NiftiMasker
+from nilearn.masking import compute_background_mask, _extrapolate_out_mask
+from nilearn.plotting import plot_stat_map
+
+
+# Use for caching results to disk, so running twice will be *fast*
+mem = Memory('cache', verbose=0)
+
+
+@mem.cache
+def splitext_multi(p):
+    """From my_img.nii.gz, return .nii.gz"""
+    dirname, basename = os.path.dirname(p), os.path.basename(p)
+    stem, ext = basename.split('.', 1)
+    return os.path.join(dirname, stem), '.' + ext if ext else ''
+
+
+@mem.cache
+def resample_and_expand_image(src_img, target_img, out_path=None):
+    """Resample src_img to target_img affine, turn 4D into list of 3D.
+
+    Optionally save to out_path.
+    """
+    data = src_img.get_data().squeeze()
+    data[np.isnan(data)] = 0
+    assert np.abs(data).max() > 0
+
+    if out_path is not None:
+        # Create the proper output filenames (if 3D, just one;
+        #   if 4D then one per time point)
+        out_stem, out_ext = splitext_multi(out_path)
+        out_paths = [out_path] if len(data.shape) == 3 else \
+                    ['%s-%d%s' % (out_stem, ii, out_ext)
+                     for ii in range(data.shape[-1])]
+        # If all output images exist, use them cached from disk.
+        if np.all([os.path.exists(p) for p in out_paths]):
+            return [nib.load(p) for p in out_paths]
+
```

Comment:
Start comment (content of the comment below, sorry, I don't see a better way of doing this)